src/windows/Button.tsx

```diff
@@ -0,0 +1,211 @@
+/**
+* Button.tsx
+*
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT license.
+*
+* RN Windows-specific implementation of the cross-platform Button abstraction.
+*/
+
+import React = require('react');
+import {Button as ButtonBase} from '../native-common/Button';
+import EventHelpers from '../native-common/utils/EventHelpers';
+import UserInterface from '../native-desktop/UserInterface';
+import RN = require('react-native');
+import RNW = require('react-native-windows');
+import { applyFocusableComponentMixin, FocusManagerFocusableComponent } from '../native-desktop/utils/FocusManager';
+
+const KEY_CODE_ENTER = 13;
+const KEY_CODE_SPACE = 32;
+
+const DOWN_KEYCODES = [KEY_CODE_SPACE, KEY_CODE_ENTER];
+const UP_KEYCODES = [KEY_CODE_SPACE];
+
+let _isNavigatingWithKeyboard = false;
+
+UserInterface.keyboardNavigationEvent.subscribe(isNavigatingWithKeyboard => {
+   _isNavigatingWithKeyboard = isNavigatingWithKeyboard;
+});
+
+// Simple check for the presence of the updated React Native for Windows
+const IsUpdatedReactNativeForWindows = (RNW.FocusableWindows !== undefined);
+
+export class Button extends ButtonBase implements FocusManagerFocusableComponent {
+
+    private _focusableElement : RNW.FocusableWindows | null = null;
+
+    private _isMouseOver = false;
+    private _isFocusedWithKeyboard = false;
+    private _isHoverStarted = false;
+
+    private _onFocusableRef = (btn: RNW.FocusableWindows | null): void => {
+        this._focusableElement = btn;
+    }
+
+    protected _render(internalProps: RN.ViewProps): JSX.Element {
+
+        // Fallback to native-common fast if the keyboard enabled component is not available
+        if (!IsUpdatedReactNativeForWindows) {
+            return super._render(internalProps);
+        }
+
+        // RNW.FocusableProps tabIndex: default is 0.
+        // -1 has no special semantic similar to DOM.
+        let tabIndex: number | undefined = this.getTabIndex();
+        // RNW.FocusableProps windowsTabFocusable:
+        // - true: keyboard focusable through any mean, receives keyboard input
+        // - false: not focusable at all, doesn't receive keyboard input
+        // The intermediate "focusable, but not in the tab order" case is not supported.
+        let windowsTabFocusable: boolean = !this.props.disabled && tabIndex !== undefined && tabIndex >= 0;
```

Comment:
>!this.props.disabled && tabIndex !== undefined && tabIndex >= 0 [](start = 43, length = 63)

Should we put this in a helper? I see we have this logic a couple of times.

Reply:
may become unclear. There are subtle differences... sometimes we have a "disabled", sometime not, etc.

---
In reply to: [159337425](https://github.com/Microsoft/reactxp/pull/404#discussion_r159337425) [](ancestors = 159337425)